src/Functions/Regexps.h

```diff
@@ -38,254 +37,283 @@ namespace ErrorCodes
+
+    if constexpr (like)
+        return {likePatternToRegexp(pattern), flags};
+    else
+        return {pattern, flags};
+}
 
-    template<bool no_capture, bool case_insensitive>
-    inline int buildRe2Flags()
+/// Caches compiled re2 objects for given string patterns. Intended to support the common situation of a small set of patterns which are
+/// evaluated over and over within the same query. In these situations, usage of the cache will save unnecessary pattern re-compilation.
+/// However, we must be careful that caching does not add too much static overhead to overall pattern evaluation. Therefore, the cache is
+/// intentionally very lightweight: a) no thread-safety/mutexes, b) small & fixed capacity, c) no collision list, d) but also no open
+/// addressing, instead collisions simply replace the existing element.
+class LocalCacheTable
+{
+public:
+    using RegexpPtr = std::shared_ptr<Regexp>;
+
+    LocalCacheTable()
+        : known_regexps(max_regexp_cache_size, {"", nullptr})
     {
-        int flags = OptimizedRegularExpression::RE_DOT_NL;
-        if constexpr (no_capture)
-            flags |= OptimizedRegularExpression::RE_NO_CAPTURE;
-        if constexpr (case_insensitive)
-            flags |= OptimizedRegularExpression::RE_CASELESS;
-        return flags;
     }
 
-    /** Returns holder of an object from Pool.
-      * You must hold the ownership while using the object.
-      * In destructor, it returns the object back to the Pool for further reuse.
-      */
     template <bool like, bool no_capture, bool case_insensitive>
-    inline Pool::Pointer get(const std::string & pattern)
+    void getOrSet(const String & pattern, RegexpPtr & regexp)
     {
-        /// the Singleton is thread-safe in C++11
-        static Pool known_regexps; /// Different variables for different pattern parameters.
+        StringAndRegexp & bucket = known_regexps[hasher(pattern) % max_regexp_cache_size];
 
-        return known_regexps.get(pattern, [&pattern]
+        if (likely(bucket.regexp != nullptr))
         {
-            const int flags = buildRe2Flags<no_capture, case_insensitive>();
-            ProfileEvents::increment(ProfileEvents::RegexpCreated);
-            return new Regexp{createRegexp<like>(pattern, flags)};
-        });
+            if (pattern == bucket.pattern)
+                regexp = bucket.regexp;
+            else
+            {
+                regexp = std::make_shared<Regexp>(createRegexp<like, no_capture, case_insensitive>(pattern));
+                bucket = {pattern, regexp};
+            }
+        }
+        else
+        {
+            regexp = std::make_shared<Regexp>(createRegexp<like, no_capture, case_insensitive>(pattern));
+            bucket = {pattern, regexp};
+        }
     }
+
+private:
+    std::hash<std::string> hasher;
+    struct StringAndRegexp
+    {
+        std::string pattern;
+        RegexpPtr regexp;
+    };
+    using CacheTable = std::vector<StringAndRegexp>;
```

Comment:
This could be a `std::array` right?

Reply:
True ... https://github.com/ClickHouse/ClickHouse/pull/37806 ... @antonio2368 May I set you as reviewer?